File: frontend/src/features/practice/__tests__/runner-machine.test.ts
import { describe, it, expect } from "vitest";
import {
  initialRunnerState,
  toggleSelection,
  canSubmit,
  markSubmitted,
} from "@/features/practice/runner-machine";
import type { AnswerResult } from "@/lib/api/types";

const result: AnswerResult = {
  is_correct: true,
  correct_indexes: [0],
  selected_indexes: [0],
  correct_rationale: { en: "because", zh: null },
  key_point_summary: { en: null, zh: null },
  per_option: [],
  mapping: {},
  history: [],
};

describe("runner machine", () => {
  it("fresh question starts selecting with no selection", () => {
    const s = initialRunnerState(null);
    expect(s.phase).toBe("selecting");
    expect(s.selected).toEqual([]);
    expect(canSubmit(s)).toBe(false);
  });

  it("rehydrates an already-answered question as submitted with its prior selection", () => {
    const s = initialRunnerState({ selected: [2], is_correct: false });
    expect(s.phase).toBe("submitted");
    expect(s.selected).toEqual([2]);
    expect(canSubmit(s)).toBe(false);
  });

  it("single_choice selection replaces the prior choice", () => {
    let s = initialRunnerState(null);
    s = toggleSelection(s, 1, "single_choice");
    s = toggleSelection(s, 3, "single_choice");
    expect(s.selected).toEqual([3]);
    expect(canSubmit(s)).toBe(true);
  });

  it("true_false selection replaces the prior choice", () => {
    let s = initialRunnerState(null);
    s = toggleSelection(s, 0, "true_false");
    s = toggleSelection(s, 1, "true_false");
    expect(s.selected).toEqual([1]);
  });

  it("multiple_choice toggles selections in and out, kept sorted", () => {
    let s = initialRunnerState(null);
    s = toggleSelection(s, 2, "multiple_choice");
    s = toggleSelection(s, 0, "multiple_choice");
    expect(s.selected).toEqual([0, 2]);
    s = toggleSelection(s, 2, "multiple_choice");
    expect(s.selected).toEqual([0]);
  });

  it("cannot toggle after submitting", () => {
    let s = initialRunnerState(null);
    s = toggleSelection(s, 1, "single_choice");
    s = markSubmitted(s, result);
    const after = toggleSelection(s, 2, "single_choice");
    expect(after.selected).toEqual([1]);
    expect(after.phase).toBe("submitted");
  });

  it("markSubmitted captures the result and locks the phase", () => {
    let s = initialRunnerState(null);
    s = toggleSelection(s, 0, "single_choice");
    s = markSubmitted(s, result);
    expect(s.phase).toBe("submitted");
    expect(s.result).toBe(result);
    expect(canSubmit(s)).toBe(false);
  });
});

File: mobile/test/runner_machine_test.dart
import 'package:cissp_api/cissp_api.dart';
import 'package:test/test.dart';

import 'package:cissp_compass/features/practice/runner_machine.dart';

AnswerResult _result({required bool correct, List<int> selected = const [0]}) {
  return AnswerResult(
    isCorrect: correct,
    correctIndexes: const [0],
    selectedIndexes: selected,
    correctRationale: const Localized(en: 'why'),
    keyPointSummary: const Localized(en: 'kp'),
    perOption: const [],
  );
}

void main() {
  group('runner_machine', () {
    test('initial state is selecting when no previous answer', () {
      final s = initialRunnerState(null);
      expect(s.phase, RunnerPhase.selecting);
      expect(s.selected, isEmpty);
      expect(s.result, isNull);
    });

    test('initial state is submitted when previous answer exists', () {
      final s = initialRunnerState(
        const PreviousAnswer(selected: [1, 2], isCorrect: false),
      );
      expect(s.phase, RunnerPhase.submitted);
      expect(s.selected, [1, 2]);
    });

    test('single-choice toggle replaces selection', () {
      var s = initialRunnerState(null);
      s = toggleSelection(s, 0, 'single_choice');
      expect(s.selected, [0]);
      s = toggleSelection(s, 2, 'single_choice');
      expect(s.selected, [2]);
    });

    test('multiple-choice toggle adds and removes', () {
      var s = initialRunnerState(null);
      s = toggleSelection(s, 1, 'multiple_choice');
      s = toggleSelection(s, 0, 'multiple_choice');
      expect(s.selected, [0, 1]);
      s = toggleSelection(s, 1, 'multiple_choice');
      expect(s.selected, [0]);
    });

    test('toggle is a no-op after submit', () {
      var s = initialRunnerState(null);
      s = toggleSelection(s, 0, 'single_choice');
      s = markSubmitted(s, _result(correct: true));
      final after = toggleSelection(s, 1, 'single_choice');
      expect(after.selected, [0]);
      expect(after.phase, RunnerPhase.submitted);
    });

    test('canSubmit requires selecting phase with non-empty selection', () {
      expect(canSubmit(initialRunnerState(null)), isFalse);
      final withSel = toggleSelection(initialRunnerState(null), 0, 'single_choice');
      expect(canSubmit(withSel), isTrue);
      final submitted = markSubmitted(withSel, _result(correct: false));
      expect(canSubmit(submitted), isFalse);
    });

    test('markSubmitted attaches result and moves to submitted', () {
      final s = toggleSelection(initialRunnerState(null), 0, 'single_choice');
      final result = _result(correct: true);
      final next = markSubmitted(s, result);
      expect(next.phase, RunnerPhase.submitted);
      expect(next.result, same(result));
      expect(next.selected, [0]);
    });
  });
}
